### DISCUSSION TITLE
Iterate over cell Range

### DISCUSSION THREAD

**Discussion by doczoidberg** [handsontable/hyperformula#522] [General]
how can I get all cell values of a cell Range?

i need to iterate over all CellAddresses of `hfInstance.getCellPrecedents`

**Comment by wojciechczerniak** [comment]
Hi @doczoidberg, would you mind sharing the solution? Was it ` getRangeValues` or you had a different use case in mind?

**Comment by doczoidberg** [comment]
i did not found a solution yet

**Comment by wojciechczerniak** [comment]
It's described in docs [`getRangeValues`](https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#getrangevalues), let us know if it's what you were looking for.

Example:
```ts
const hfInstance = HyperFormula.buildFromArray([
 ['=SUM(1,2)', '2', '10'],
 ['5', '6', '7'],
 ['40', '30', '20'],
]);

// returns calculated cells content: [ [ 3, 2 ], [ 5, 6 ] ]
const rangeValues = hfInstance.getRangeValues({ sheet: 0, col: 0, row: 0 }, 2, 2);

```

**Comment by doczoidberg** [comment]
but in your exmplae the parameter is not a cell range

**Comment by doczoidberg** [comment]
i need to iterate over all CellAddresses of `hfInstance.getCellPrecedents`

**Comment by doczoidberg** [comment]
I try 
`    console.log((this.hfInstance.getCellPrecedents(this.hfInstance.simpleCellAddressFromString(cell, sheet)) as any).arrayOfAddressesInRange());
`
but get 

`.arrayOfAddressesInRange is not a function`

**Comment by doczoidberg** [comment]
solution:

`(this.hfInstance.getCellPrecedents(this.hfInstance.simpleCellAddressFromString(cell, sheet)) as unknown as AbsoluteCellRange)[0].arrayOfAddressesInRange()`

**Comment by wojciechczerniak** [comment]
> but in your exmplae the parameter is not a cell range

It is a Range starting in cell (0,0) and spreading across 2 rows and 2 columns. With `CellRange` object you would use `cellRange.start`, `cellRange.width` and `cellRange.height` to call it. ~But I don't remember any public API that would return it.~ Ok, I see it now.

> i need to iterate over all CellAddresses of `hfInstance.getCellPrecedents`

`getCellPrecedents` does not return a Range. They're not necessarily contained within a rectangle, those cells may be in different Sheets. In that case, I would go with simple `getCellValue` in a loop.

```ts
const precedents = hfInstance.getCellPrecedents();

const precedentsValues = precedents.map((simpleCellAddress) => hfInstance.getCellValue(simpleCellAddress));
```

**Comment by wojciechczerniak** [comment]
Found it. You're right. My solution has to be extended with support for the `AbsoluteRange` object.

```ts
const precedents = hfInstance.getCellPrecedents();

const precedentsValues = precedents.map((precedent) => {
  if (precedent.start) {
    return hfInstance.getRangeValues(precedent.start, precedent.width, precedent.height);
  }

  return hfInstance.getCellValue(simpleCellAddress);
});
```

@izulin shouldn't we have an `AbsoluteRange` export for more type friendly check?
